File: 198-house-robber/house-robber.java
class Solution {
    /* the order is: prev2, prev1, num  */
int[] memo;
public int rob(int[] nums) {
    memo = new int[nums.length + 1];
    Arrays.fill(memo, -1);
    return rob(nums, nums.length - 1);
}

private int rob(int[] nums, int i) {
    if (i < 0) {
        return 0;
    }
    if (memo[i] >= 0) {
        return memo[i];
    }
    int result = Math.max(rob(nums, i - 2) + nums[i], rob(nums, i - 1));
    memo[i] = result;
    return result;
}
}

File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int prevRob = 0;
        int currRob = 0;
        for (int house : nums) {
            int tmp = currRob;
            currRob = max(prevRob + house, currRob);
            prevRob = tmp;
        }
        return currRob;
    }
};